pkg/containerruntime/containerruntime.go

```diff
@@ -0,0 +1,97 @@
+/*
+Copyright 2020 The Machine Controller Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package containerruntime
+
+import (
+	"github.com/Masterminds/semver"
+
+	"github.com/kubermatic/machine-controller/pkg/providerconfig/types"
+)
+
+const (
+	Default = "docker"
+)
+
+type Engine interface {
+	KubeletFlags() []string
+	ScriptFor(os types.OperatingSystem) (string, error)
+	ConfigFileName() string
+	Config() (string, error)
+}
+
+func Get(containerRuntimeName string) Config {
+	cfg := Config{}
+
+	switch containerRuntimeName {
+	case "docker":
+		cfg.Docker = &Docker{}
+		cfg.Containerd = nil
+	case "containerd":
+		cfg.Containerd = &Containerd{}
+		cfg.Docker = nil
+	default:
+		cfg.Docker = &Docker{}
+		cfg.Containerd = nil
+	}
+
+	return cfg
+}
+
+type Config struct {
+	Docker             *Docker     `json:",omitempty"`
+	Containerd         *Containerd `json:",omitempty"`
+	InsecureRegistries []string    `json:",omitempty"`
+	RegistryMirrors    []string    `json:",omitempty"`
+}
+
+func (cfg Config) String() string {
+	switch {
+	case cfg.Containerd != nil:
+		return "containerd"
+	case cfg.Docker != nil:
+		return "docker"
+	}
+
+	return Default
+}
+
+func (cfg Config) Engine(kubeletVersion *semver.Version) Engine {
+	var (
+		docker = &Docker{
+			insecureRegistries: cfg.InsecureRegistries,
+			registryMirrors:    cfg.RegistryMirrors,
+			kubeletVersion:     kubeletVersion,
+		}
+		containerd = &Containerd{
+			insecureRegistries: cfg.InsecureRegistries,
+			registryMirrors:    cfg.RegistryMirrors,
+		}
+	)
+
+	moreThen123, _ := semver.NewConstraint(">= 1.23")
```

Comment:
This is the supposed version where dockershim will be removed. But it's unknown for real. If and hen there will be announced another version to remove dockershim, we will change it here.

Reply:
As per the dockershim removal KEP, this should be 1.22.